### ISSUE TITLE
Hydrator generation fails silently when using custom types

### ISSUE THREAD

**Issue by melkamar** [doctrine/mongodb-odm#1889]
### Bug Report

Only some Hydrator classes are generated when using a custom type. Once the HydratorFactory tries to generate a Hydrator for a custom type, it fails silently and just stops generating the rest. Subsequent tests then fail at a `require` which cannot find the Hydrator file it is expecting.

|    Q        |   A
|------------ | ------
| BC Break    | no
| Version     | 1.2.5

#### Summary

This only happens when generating as a part of cache warmup, but succeeds when running the doctrine command to generate hydrators, i.e. 

```bash
# THIS FAILS
$ app/console cache:clear -e test
$ ./vendor/bin/phpunit -c app/phpunit.ci.xml --filter testCreateProductWithNumericPrice src

# THIS SUCCEEDS
$ app/console cache:clear -e test
$ app/console doctrine:mongodb:generate:hydrators -e test
$ ./vendor/bin/phpunit -c app/phpunit.ci.xml --filter testCreateProductWithNumericPrice src
```

I have stepped through and made some debugging output from the HydratorFactory and I can see that in the first case Hydrators are actually being created - the code is being run and is expected to create 203 classes. However, it only generates about 80, after which a hydrator for my custom class is expected to be created and the execution stops, with no errors logged.

The actual error when running PHPUnit is:
```
./vendor/bin/phpunit -c app/phpunit.ci.xml --exclude-group stats,integration,heavy --log-junit /tmp/phpunit_test_results.xml src/
PHP Warning:  require(/home/example/platform/app/cache/test/doctrine/odm/mongodb/Hydrators/RHLocationBundleDocumentLocationHydrator.php): failed to open stream: "VCR\Util\StreamProcessor::stream_open" call failed in /home/example/platform/vendor/doctrine/mongodb-odm/lib/Doctrine/ODM/MongoDB/Hydrator/HydratorFactory.php on line 142
PHP Fatal error:  require(): Failed opening required '/home/example/platform/app/cache/test/doctrine/odm/mongodb/Hydrators/RHLocationBundleDocumentLocationHydrator.php' (include_path='.:/usr/share/php') in /home/example/platform/vendor/doctrine/mongodb-odm/lib/Doctrine/ODM/MongoDB/Hydrator/HydratorFactory.php on line 142
PHP Fatal error:  Uncaught Error: Class 'Symfony\Component\Debug\Exception\FatalErrorException' not found in /home/example/platform/vendor/symfony/symfony/src/Symfony/Component/Debug/ErrorHandler.php:670
Stack trace:
#0 [internal function]: Symfony\Component\Debug\ErrorHandler::handleFatalError()
#1 {main}
  thrown in /home/example/platform/vendor/symfony/symfony/src/Symfony/Component/Debug/ErrorHandler.php on line 670
Script ./vendor/bin/phpunit -c app/phpunit.ci.xml --exclude-group stats,integration,heavy --log-junit /tmp/phpunit_test_results.xml src/ handling the test-ci event returned with error code 255
```

The custom class implementation is this:
```php
use Doctrine\ODM\MongoDB\Types\Type;

/**
 * Custom type for storing instances of \Money\Money in MongoDB.
 */
class MoneyMongoType extends Type
{
    public function convertToDatabaseValue($value)
    {
        return [
            'amount' => $value->getAmount(),
            'currency' => $value->getCurrency()->getCode()
        ];
    }

    public function convertToPHPValue($value)
    {
        return new \Money\Money($value['amount'], new \Money\Currency($value['currency']));
    }

    public function closureToPHP()
    {
        return '$return = new \Money\Money($value[\'amount\'], new \Money\Currency($value[\'currency\']));';
    }

}
```

Everything works as long as I do not include my custom type in any `Entity.mongodb.yml` configuration - i.e. as long as no Hydrator has to be generated for that custom type.

#### Expected behavior

I should not need to run the command to generate hydrators manually, when the generation is executed after clearing the cache anyway.

**Comment by malarzm** [comment]
> I have stepped through and made some debugging output from the HydratorFactory and I can see that in the first case Hydrators are actually being created - the code is being run and is expected to create 203 classes. However, it only generates about 80, after which a hydrator for my custom class is expected to be created and the execution stops, with no errors logged.

Any chance you could step over generation process and see exactly what is failing? My first idea is that somehow custom types are not registered during cache warmup, but that's a blind guess.

**Comment by melkamar** [comment]
Sorry for the long wait. I have tried stepping through it when I encountered the issue, but that was quite hard because no exception was raised. What I know for sure is that this happened in `HydratorFactory::generateHydratorClass()`. That leads me to believe that the custom type was registered, but somehow the generation failed.

This was a feature we needed to ship so I worked around it and didn't end up using a custom type. I will try and see if I can reproduce if I have some time in the near future.
